neo/Consensus/ConsensusService.cs

```diff
@@ -225,18 +273,273 @@ private void OnPersistCompleted(Block block)
+
+            UInt256 preparationHash;
+            if (message.PreparationMessages.ContainsKey((ushort)tempContext.PrimaryIndex)
+                && ReverifyPrepareRequest(tempContext, message, snap, out prepareRequestPayload, out prepareRequestMessage))
+                preparationHash = prepareRequestPayload.Hash;
+            else
+                preparationHash = message.PreparationHash;
+
+            var prepareResponses = new List<(ConsensusPayload, PrepareResponse)>();
+            var prepareResponseMsg = new PrepareResponse
+            {
+                ViewNumber = tempContext.ViewNumber,
+                PreparationHash = preparationHash
+            };
+            bool canRestoreView = false;
+            if (preparationHash != null)
+            {
+                for (ushort i = 0; i < context.Validators.Length; i++)
+                {
+                    if (i == tempContext.PrimaryIndex) continue;
+                    if (!message.PreparationMessages.ContainsKey(i)) continue;
+                    Log($" considering prepare request {i}");
+                    var regeneratedPrepareResponse = tempContext.RegenerateSignedPayload(prepareResponseMsg, i,
+                        message.PreparationMessages[i].InvocationScript);
+                    if (!regeneratedPrepareResponse.Verify(snap) ||
+                        !PerformBasicConsensusPayloadPreChecks(regeneratedPrepareResponse)) continue;
+                    prepareResponses.Add((regeneratedPrepareResponse, prepareResponseMsg));
+                    Log($" verified prepare {i}");
+                    // Verify that there are M valid preparations, 1 Prepare Request + (M-1) Prepare responses
+                    if (prepareRequestMessage == null || prepareResponses.Count < context.M - 1) continue;
+                    canRestoreView = true;
+                    break;
+                }
+            }
+
+            // Only accept recovery from lower views if there were at least M valid prepare requests
+            if (message.ViewNumber < context.ViewNumber && !canRestoreView) return;
+
+            var verifiedChangeViewPayloads = new ConsensusPayload[context.Validators.Length];
+            if (!canRestoreView && message.ChangeViewMessages.Count >= context.M)
+            {
+                var changeViewMsg = new ChangeView { NewViewNumber = message.ViewNumber };
+                int validChangeViewCount = 0;
+                for (ushort i = 0; i < context.Validators.Length; i++)
+                {
+                    if (!message.ChangeViewMessages.ContainsKey(i)) continue;
+
+                    changeViewMsg.ViewNumber = message.ChangeViewMessages[i].OriginalViewNumber;
+                    changeViewMsg.Timestamp = message.ChangeViewMessages[i].Timestamp;
+                    // Regenerate the ChangeView message
+                    var regeneratedChangeView = tempContext.RegenerateSignedPayload(changeViewMsg, (ushort) i,
```

Comment:
If `i` is declared in the loops as a ushort this doesn’t need to cast. It seems like it is better to make it defines as int though and leave the cast.